provider/formtype_test.go

```diff
@@ -0,0 +1,426 @@
+					formType:   ft,
+					optionType: ot,
+					options:    true,
+				})
+			}
+		}
+
+		for _, check := range requiredChecks {
+			_, alreadyChecked := formTypesChecked[check.String()]
+			if alreadyChecked {
+				continue
+			}
+
+			ftName := check.formType
+			if ftName == "" {
+				ftName = "default"
+			}
+			fc := formTypeTestCase{
+				name: fmt.Sprintf("%s_%s_%t",
+					ftName,
+					check.optionType,
+					check.options,
+				),
+				config:      check,
+				assert:      paramAssert{},
+				expectError: regexp.MustCompile("is not supported"),
+			}
+
+			t.Run(fc.name, func(t *testing.T) {
+				t.Parallel()
+
+				// This is just helpful log output to give the boilerplate
+				// to write the manual test.
+				tcText := fmt.Sprintf(`
+					obvious(%s, ezconfigOpts{
+						Options:    %t,
+						OptionType: %q,
+						FormType:   %q,
+					}),
+				//`, "<expected_form_type>", check.options, check.optionType, check.formType)
+
+				probablyPassed := formTypeTest(t, fc)
+				if !probablyPassed {
+					t.Logf("To construct this test case:\n%s", tcText)
+				}
+			})
+
+		}
+	})
+}
+
+// ezconfig converts a formTypeCheck into a terraform config string.
+func ezconfig(paramName string, cfg formTypeCheck) (defaultValue string, tf string) {
+	options := cfg.customOptions
+	if cfg.options && len(cfg.customOptions) == 0 {
+		switch cfg.optionType {
+		case provider.OptionTypeString:
+			options = []string{"foo"}
+			defaultValue = "foo"
+		case provider.OptionTypeBoolean:
+			options = []string{"true", "false"}
+			defaultValue = "true"
+		case provider.OptionTypeNumber:
+			options = []string{"1"}
+			defaultValue = "1"
+		case provider.OptionTypeListString:
+			options = []string{`["red", "blue"]`}
+			defaultValue = `["red", "blue"]`
+		default:
+			panic(fmt.Sprintf("unknown option type %q when generating options", cfg.optionType))
+		}
+	}
+
+	if cfg.defValue == "" {
+		cfg.defValue = defaultValue
+	}
+
+	var body strings.Builder
+	if cfg.defValue != "" {
+		body.WriteString(fmt.Sprintf("default = %q\n", cfg.defValue))
+	}
+	if cfg.formType != "" {
+		body.WriteString(fmt.Sprintf("form_type = %q\n", cfg.formType))
+	}
+	if cfg.optionType != "" {
+		body.WriteString(fmt.Sprintf("type = %q\n", cfg.optionType))
+	}
+	if cfg.styling != nil {
+		body.WriteString(fmt.Sprintf("styling = %s\n", strconv.Quote(string(cfg.styling))))
+	}
+
+	for i, opt := range options {
+		body.WriteString("option {\n")
+		body.WriteString(fmt.Sprintf("name = \"val_%d\"\n", i))
+		body.WriteString(fmt.Sprintf("value = %q\n", opt))
+		body.WriteString("}\n")
+	}
+
+	return cfg.defValue, fmt.Sprintf(`
+				provider "coder" {
+				}
+				data "coder_parameter" "%s" {
+					name = "%s"
+					%s
+				}
+		`, paramName, paramName, body.String())
+}
+
+func formTypeTest(t *testing.T, c formTypeTestCase) bool {
+	t.Helper()
+	const paramName = "test_param"
+	// probablyPassed is just a guess used for logging. It's not important.
+	probablyPassed := true
```

Comment:
One the one hand, you say that it's not important, but it's returned by the function here. This implies to me that it is arbitrarily as important as whatever callers decide to do with it?

Reply:
If I rename it to `logDebugInfo`, then it's more accurate to what it is doing.